docs/rfcs/2024-04-01-super-peers-2.md

```diff
@@ -0,0 +1,262 @@
+
+The decision whether a given super peer should deliver the message would depend on these factors:
+- deterministic (see below) message hash, `h`.
+- receiving member ID, `r`.
+- super-peer member ID, `s`.
+- target message redundancy `d` (the desired number of super-peers to deliver the message).
+- number of super-peers connected to member `r` (and known as connected to super-peer making the decision), `n`.
+- 0-based index of the current super-peer in the sorted array of known super peer IDs, `i`. It does not require sorting all peer IDs, it's enough to count how many peers have smaller or larger IDs.
+
+The assumption here is that member ID is unique within the group, and that message hashes are also unique. Also, message hashes rather than sending member IDs are used to ensure that the decision is made differently for the same sender/recipient pairs, allowing recipients to identify integrity violations (e.g., if some of the super-peers decides to change the messages or fail to deliver it).
+
+For simplicity, all parameters are normalized to 0..1 range from their respective binary ranges.
+
+The algorithm to decide whether the message should be delivered by a given super-peer:
+
+```
+if (d >= n || n == 1) deliver;
+else
+  prob = d/n ; delivery probability based on target delivery redundancy
+  point = (h + r)/2 ; as member ID and message hash are uniformly random in 0..1 range, `point` will also be uniformly random in 0..1 range
+  start = (1 - prob) * i / (n - 1) ; `start` for the peer with `i == 0` will be `0` and with `i == n - 1` will be `1 - prob`
+  end = start + prob ; `end` for peer 0 will be `prob` and for peer `n - 1` will be 1
+  ; for all peers the range `start..end` will have width `prob`
+  if (point >= start && point < end) deliver
+  else skip;
+```
+
+This algorithm can be proven to result in target average delivery redundancy.
+
+It also requires all super peers to inform other super peers about:
+- establishing or losing the connection with other members (e.g., when AUTH is received).
+- informing about their decision to stop being super peers in advance.
+- most likely using delivery receipts in communications between super-peers that would include message hashes in RCVD info.
+
+## Authorising administrative changes
+
+As members no longer send messages directly to other members, in addition to the risk of the initial MITM by admin (which is now mitigated by having multiple super-peers) there is a risk of super-peers being compromised at a later stage, and in case of administrative changes (member or group level changes) we have these options:
+
+1. have all members postpone these changes until they are communicated by a sufficient number of super-peers (that would require configuring consensus on the group level).
+2. have admins and owners sign administrative changes with the public key included in their profile during member introduction.
```

Comment:
which is also passed by super-peers? if so, also adds to the point that being a super-peer requires a certain level of trust

Reply:
Yes, but it protects from future compromise. Plus super-peer redundancy validates the keys - they can be controlled by different parties.